In [1]:
import pandas as pd

In [108]:
data = pd.read_csv("./student.csv")

In [109]:
#แสดงข้อมูล
data

รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
2          64003       NaN    NaN    Male         Science  3.67       175.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
10         64011      ธีระ    NaN     NaN             NaN  3.78         NaN   
11         64012       นภา   22.0     ชาย           วิทย์   NaN       171.0   
12         64013     บุญมี   20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0     ชาย    Liberal Arts  3.45       164.0   
14         64015     พิมพ์   22.0  Female             Sci  3.67       162.0   
15         64016      ภูมิ   19.0       ช     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0       ญ  วิศวกรรมศาสตร์  3.01       166.0   
17         64018      ยุพา   21.0    หญิง             NaN  3.23       161.0   
18         64019     รัตนา    NaN     ชาย      ศิลปศาสตร์   NaN       174.0   
19         64020     ลัดดา   22.0     NaN         Science  3.45       168.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
2          NaN                     25.0      2024-01-13  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
10        68.0                      NaN      2024-01-07  
11         NaN                     16.0      2023-12-26  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24  
17        50.0                     27.0      2024-01-08  
18        76.0                      NaN      2023-12-18  
19        64.0                     23.0      2024-01-06

In [111]:
data.shape

(14, 10)

# **เช็คค่าที่ผิดปกติในแต่ละคอลัม**

In [112]:

print("="*50)
print("🎯 ข้อมูลหลัก (Data) ที่จะใช้ตลอดบทเรียน")
print("="*50)
print(data)
print("\n📊 สรุปปัญหาที่พบในข้อมูล:")
print(f"- ค่าหายทั้งหมด: {data.isnull().sum().sum()} ค่า")
print(f"- อายุผิดปกติ: {data[data['อายุ'] > 100]['อายุ'].values if any(data['อายุ'] > 100) else 'ไม่มี'}")
print(f"- GPA ผิดปกติ: {data[(data['GPA'] < 0) | (data['GPA'] > 4)]['GPA'].values if any((data['GPA'] < 0) | (data['GPA'] > 4)) else 'ไม่มี'}")
print(f"- รูปแบบการเขียนคณะ: {data['คณะ'].nunique()} แบบ")
print(f"- รูปแบบการเขียนเพศ: {data['เพศ'].nunique()} แบบ")

🎯 ข้อมูลหลัก (Data) ที่จะใช้ตลอดบทเรียน
    รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  20.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี  20.0  หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยา

# **คำสั่งในการเก็บสำเนาเพื่อไม่ให้กระทบกับข้อมูลต้นฉบับ**

In [73]:
df_original = data.copy()

# **ตรวจสอบค่าหายในแต่ละคอลัมน์**

In [74]:
# ตรวจสอบค่าหายในแต่ละคอลัมน์
missing_counts = data.isnull().sum()
print("\n❌ ค่าหายในแต่ละคอลัมน์:")
for col, count in missing_counts[missing_counts > 0].items():
    percentage = (count / len(data)) * 100
    print(f"   - {col}: {count} ค่า ({percentage:.1f}%)")


❌ ค่าหายในแต่ละคอลัมน์:
   - ชื่อ: 1 ค่า (5.0%)
   - อายุ: 3 ค่า (15.0%)
   - เพศ: 2 ค่า (10.0%)
   - คณะ: 2 ค่า (10.0%)
   - GPA: 2 ค่า (10.0%)
   - ส่วนสูง_cm: 1 ค่า (5.0%)
   - น้ำหนัก_kg: 2 ค่า (10.0%)
   - ชั่วโมงเรียน_ต่อสัปดาห์: 2 ค่า (10.0%)


# จัดการข้อมูลหายโดยการลบแถวที่มีค่าหาย เพราะจะได้ทำข้อมูลที่ปกติและนำไปใช้ได้

In [75]:
data_clean = data.dropna()
data_clean

รหัสนักศึกษา      ชื่อ   อายุ     เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0     ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0    หญิง           วิทย์  3.45       165.0   
3          64004  สมศักดิ์   22.0       ช  คณะวิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ   19.0       ญ           วิศวะ  3.89       160.0   
5          64006     กานต์  150.0     ชาย     Engineering  3.12       172.0   
6          64007     กิตติ   21.0    หญิง        วิศวกรรม  4.50       158.0   
7          64008      จิรา   20.0       F      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0       M            Arts  3.34       163.0   
9          64010      ดารา   21.0    หญิง           ศิลป์  3.56       167.0   
12         64013     บุญมี   20.0    หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0     ชาย    Liberal Arts  3.45       164.0   
14         64015     พิมพ์   22.0  Female             Sci  3.67       162.0   
15         64016      ภูมิ   19.0       ช     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0       ญ  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

# ถ้ามีข้อมูลที่เยอะ จะใช้วิธีการเติมค่าด้วยค่าเฉลี่ยแทน ซึ่งเหมาะกับข้อมูลตัวเลข

In [76]:
mean_age = data['อายุ'].mean()
data.fillna({'อายุ': mean_age}, inplace=True)
data

รหัสนักศึกษา      ชื่อ        อายุ     เพศ             คณะ   GPA  \
0          64001     สมชาย   20.000000     ชาย     วิทยาศาสตร์  3.21   
1          64002    สมหญิง   21.000000    หญิง           วิทย์  3.45   
2          64003       NaN   28.470588    Male         Science  3.67   
3          64004  สมศักดิ์   22.000000       ช  คณะวิทยาศาสตร์ -1.00   
4          64005      สมใจ   19.000000       ญ           วิศวะ  3.89   
5          64006     กานต์  150.000000     ชาย     Engineering  3.12   
6          64007     กิตติ   21.000000    หญิง        วิศวกรรม  4.50   
7          64008      จิรา   20.000000       F      ศิลปศาสตร์  2.98   
8          64009    ชนาธิป   23.000000       M            Arts  3.34   
9          64010      ดารา   21.000000    หญิง           ศิลป์  3.56   
10         64011      ธีระ   28.470588     NaN             NaN  3.78   
11         64012       นภา   22.000000     ชาย           วิทย์   NaN   
12         64013     บุญมี   20.000000    หญิง           วิศวฯ  3.23   
13         64014     ปรีชา   21.000000     ชาย    Liberal Arts  3.45   
14         64015     พิมพ์   22.000000  Female             Sci  3.67   
15         64016      ภูมิ   19.000000       ช     วิทยาศาสตร์  3.89   
16         64017      มานี   20.000000       ญ  วิศวกรรมศาสตร์  3.01   
17         64018      ยุพา   21.000000    หญิง             NaN  3.23   
18         64019     รัตนา   28.470588     ชาย      ศิลปศาสตร์   NaN   
19         64020     ลัดดา   22.000000     NaN         Science  3.45   

    ส่วนสูง_cm  น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0        170.0        65.0                     15.0      2024-01-23  
1        165.0        58.0                     20.0      2024-02-05  
2        175.0         NaN                     25.0      2024-01-13  
3        168.0        70.0                     10.0      2023-12-30  
4        160.0        52.0                     30.0      2024-01-27  
5        172.0        75.0                     12.0      2023-12-23  
6        158.0        48.0                     35.0      2024-01-05  
7        180.0        82.0                     18.0      2024-01-23  
8        163.0        55.0                     22.0      2024-02-11  
9        167.0        60.0                     28.0      2024-01-03  
10         NaN        68.0                      NaN      2024-01-07  
11       171.0         NaN                     16.0      2023-12-26  
12       169.0        72.0                     24.0      2023-12-26  
13       164.0        54.0                     19.0      2024-01-08  
14       162.0        56.0                     26.0      2024-02-06  
15       173.0        78.0                     14.0      2024-01-20  
16       166.0        62.0                     21.0      2024-01-24  
17       161.0        50.0                     27.0      2024-01-08  
18       174.0        76.0                      NaN      2023-12-18  
19       168.0        64.0                     23.0      2024-01-06

In [77]:
data =data.dropna()

# **การทำให้ข้อมูลเป็นมาตรฐาน เพราะบางครั้งเราอาจพิมว่าวิทย์แต่เราหมายถึงคณะวิทยาศาสตร์**

In [116]:
def standardize_faculty(name):
    if pd.isna(name):
        return name  # ถ้าเป็น NaN ก็คืนค่า NaN ไปเลย

    name = str(name).strip()  # ตัดช่องว่างหน้า-หลัง

    # กลุ่มวิทยาศาสตร์
    if name in ['วิทยาศาสตร์', 'วิทย์', 'Science', 'Sci', 'คณะวิทยาศาสตร์']:
        return 'วิทยาศาสตร์'

    # กลุ่มศิลปศาสตร์
    elif name in ['ศิลปศาสตร์', 'ศิลป์', 'Arts', 'Liberal Arts']:
        return 'ศิลปศาสตร์'

    # กลุ่มวิศวกรรมศาสตร์
    elif name in ['วิศวะ', 'วิศวกรรม', 'วิศวกรรมศาสตร์', 'Engineering','วิศวฯ']:
        return 'วิศวกรรมศาสตร์'

    # อื่น ๆ ให้คืนค่าตามเดิม
    else:
        return name

# ใช้งานกับ DataFrame
data['คณะ'] = data['คณะ'].apply(standardize_faculty)


/tmp/ipython-input-1230001061.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['คณะ'] = data['คณะ'].apply(standardize_faculty)


# ทำให้เพศเป็นมาตราฐานเดียวกันด้วย

In [117]:
def standardize_gender(g):
    if pd.isna(g):
        return g  # ถ้าเป็น NaN ก็คืน NaN

    g = str(g).strip().lower()  # ตัดช่องว่างและแปลงเป็นตัวพิมพ์เล็ก

    # กลุ่มชาย
    if g in ['ชาย', 'ช', 'male', 'm']:
        return 'ชาย'

    # กลุ่มหญิง
    elif g in ['หญิง', 'ญ', 'female', 'f']:
        return 'หญิง'

    # อื่น ๆ คืนตามเดิม (กันกรณีมีเพศพิเศษ)
    else:
        return g

# ใช้กับ DataFrame
data['เพศ'] = data['เพศ'].apply(standardize_gender)


/tmp/ipython-input-2187384011.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['เพศ'] = data['เพศ'].apply(standardize_gender)


In [118]:
data

รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  20.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

# วิธีการจัดการค่าผิดปกติ ตัวอย่างเช่น  GPA: -1.00 หรือ อายุ150

# จัดการกับอายุที่ผิดปกติ

In [81]:
#ใช้ IQR (Interquartile Range)
Q1 = data['อายุ'].quantile(0.25)
Q3 = data['อายุ'].quantile(0.75)
IQR = Q3 - Q1

# กำหนดขอบเขต
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# กรองค่าผิดปกติ
df_no_outliers = data[(data['อายุ'] >= lower_bound) & (data['อายุ'] <= upper_bound)]

In [82]:
df_no_outliers

รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์ -1.00       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  4.50       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี  20.0  หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

# จัดการกับGPAที่ติดลบเช่น -1.00 โดยให้มันกลายเป็นค่าเฉลี่ย


In [100]:
mean_gpa = data['GPA'].mean()
data['GPA'] = data['GPA'].replace(-1.00, mean_gpa)
data['GPA'] = data['GPA'].round(2)
data

/tmp/ipython-input-2943228749.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['GPA'] = data['GPA'].replace(-1.00, mean_gpa)
/tmp/ipython-input-2943228749.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['GPA'] = data['GPA'].round(2)


รหัสนักศึกษา      ชื่อ   อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์   22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ   19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  150.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ   21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา   20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา   21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี   20.0  หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์   22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ   19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

**จัดการกับGPAที่ติดลบเช่น GPAมากกว่า4 โดยการตั้งเงื่อนไข** เพื่อไม่ให้เกิดข้อมูลที่ผิดพลาด

In [101]:
mean_gpa = data['GPA'].mean()
data.loc[ (data['GPA'] > 4.0), 'GPA'] = mean_gpa
data

รหัสนักศึกษา      ชื่อ   อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย   20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง   21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์   22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ   19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  150.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ   21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา   20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป   23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา   21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี   20.0  หญิง           วิศวฯ  3.23       169.0   
13         64014     ปรีชา   21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์   22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ   19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี   20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

แก้ไขอายุที่เป็นค่าผิดปกติ (เช่น 150.0) ให้เป็นฐานนิยม

In [120]:
mode_age = data['อายุ'].mode()[0]
data['อายุ'] = data['อายุ'].replace({150.0: mode_age})
data['GPA'] = data['GPA'].round(2)
data

/tmp/ipython-input-1253879308.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['อายุ'] = data['อายุ'].replace({150.0: mode_age})
/tmp/ipython-input-1253879308.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['GPA'] = data['GPA'].round(2)


รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  20.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยาศาสตร์  3.67       162.0   
15         64016      ภูมิ  19.0   ชาย     วิทยาศาสตร์  3.89       173.0   
16         64017      มานี  20.0  หญิง  วิศวกรรมศาสตร์  3.01       166.0   

    น้ำหนัก_kg  ชั่วโมงเรียน_ต่อสัปดาห์ วันที่ลงทะเบียน  
0         65.0                     15.0      2024-01-23  
1         58.0                     20.0      2024-02-05  
3         70.0                     10.0      2023-12-30  
4         52.0                     30.0      2024-01-27  
5         75.0                     12.0      2023-12-23  
6         48.0                     35.0      2024-01-05  
7         82.0                     18.0      2024-01-23  
8         55.0                     22.0      2024-02-11  
9         60.0                     28.0      2024-01-03  
12        72.0                     24.0      2023-12-26  
13        54.0                     19.0      2024-01-08  
14        56.0                     26.0      2024-02-06  
15        78.0                     14.0      2024-01-20  
16        62.0                     21.0      2024-01-24

# ตรวจสอบข้อมูลทั้งหมดหลังแก้ไขค่าแล้ว

In [119]:

print("="*50)
print("🎯 ข้อมูลหลัก (Data) ที่จะใช้ตลอดบทเรียน")
print("="*50)
print(data)
print("\n📊 สรุปปัญหาที่พบในข้อมูล:")
print(f"- ค่าหายทั้งหมด: {data.isnull().sum().sum()} ค่า")
print(f"- อายุผิดปกติ: {data[data['อายุ'] > 100]['อายุ'].values if any(data['อายุ'] > 100) else 'ไม่มี'}")
print(f"- GPA ผิดปกติ: {data[(data['GPA'] < 0) | (data['GPA'] > 4)]['GPA'].values if any((data['GPA'] < 0) | (data['GPA'] > 4)) else 'ไม่มี'}")
print(f"- รูปแบบการเขียนคณะ: {data['คณะ'].nunique()} แบบ")
print(f"- รูปแบบการเขียนเพศ: {data['เพศ'].nunique()} แบบ")

🎯 ข้อมูลหลัก (Data) ที่จะใช้ตลอดบทเรียน
    รหัสนักศึกษา      ชื่อ  อายุ   เพศ             คณะ   GPA  ส่วนสูง_cm  \
0          64001     สมชาย  20.0   ชาย     วิทยาศาสตร์  3.21       170.0   
1          64002    สมหญิง  21.0  หญิง     วิทยาศาสตร์  3.45       165.0   
3          64004  สมศักดิ์  22.0   ชาย     วิทยาศาสตร์  3.16       168.0   
4          64005      สมใจ  19.0  หญิง  วิศวกรรมศาสตร์  3.89       160.0   
5          64006     กานต์  20.0   ชาย  วิศวกรรมศาสตร์  3.12       172.0   
6          64007     กิตติ  21.0  หญิง  วิศวกรรมศาสตร์  3.46       158.0   
7          64008      จิรา  20.0  หญิง      ศิลปศาสตร์  2.98       180.0   
8          64009    ชนาธิป  23.0   ชาย      ศิลปศาสตร์  3.34       163.0   
9          64010      ดารา  21.0  หญิง      ศิลปศาสตร์  3.56       167.0   
12         64013     บุญมี  20.0  หญิง  วิศวกรรมศาสตร์  3.23       169.0   
13         64014     ปรีชา  21.0   ชาย      ศิลปศาสตร์  3.45       164.0   
14         64015     พิมพ์  22.0  หญิง     วิทยา